man/faq/developer/howto-coercion.Rmd

```diff
@@ -0,0 +1,222 @@
+
+```{r, child = "../setup.Rmd", include = FALSE}
+```
+
+This is a practical example of implementation of `vec_ptype2()` and `vec_cast()` methods. Related topics:
+
+- For an overview of how these generics work and their roles in vctrs, see [`?theory-faq-coercion`][theory-faq-coercion].
+
+- For an example of implementing coercion methods for data frame subclasses, see [`?howto-faq-coercion-data-frame`][howto-faq-coercion-data-frame].
+
+
+## The natural number class
+
+We'll illustrate how to implement coercion methods with a simple class that represents natural numbers. This is an existing class that features a constructor, methods for `print()` and subset.
+
+```{r}
+#' @export
+new_natural <- function(x) {
+  if (is.numeric(x) || is.logical(x)) {
+    stopifnot(is_whole(x))
+    x <- as.integer(x)
+  } else {
+    stop("Can't construct natural from unknown type.")
+  }
+  structure(x, class = "my_natural")
+}
+is_whole <- function(x) {
+  all(x %% 1 == 0 | is.na(x))
+}
+
+#' @export
+print.my_natural <- function(x, ...) {
+  cat("<natural>\n")
+  x <- unclass(x)
+  NextMethod()
+}
+#' @export
+`[.my_natural` <- function(x, i, ...) {
+  new_natural(NextMethod())
+}
+```
+
+```{r}
+new_natural(1:3)
+new_natural(c(1, NA))
+```
+
+
+## Implementing `vec_ptype2()`
+
+Implementing double dispatch methods is very similar to implementing regular S3 methods. A typical signature for a `vec_ptype2()` methods looks like this. In these examples we are using roxygen2 tags to register the methods, but you can also register the methods manually in your NAMESPACE file or lazily with `s3_register()`.
+
+
+### The self-self method
+
+The first method to implement is the one that signals that your class is compatible with itself:
+
+```{r}
+#' @export
+vec_ptype2.my_natural.my_natural <- function(x, y, ...) {
+  x
+}
+
+vec_ptype2(new_natural(1), new_natural(2:3))
+```
+
+`vec_ptype2()` implements fallbacks to try and be compatible with simple classes, so it may seem that you don't need to implement the self-self coercion methods. However, you still have to implement this method implictly because this is how vctrs knows that it is dealing with a class that implements vctrs methods. Also, this is a bit more efficient.
```

Comment:
```suggestion
`vec_ptype2()` implements fallbacks to try and be compatible with simple classes, so it may seem that you don't need to implement the self-self coercion methods. However, you still have to implement this method implicitly because this is how vctrs knows that it is dealing with a class that implements vctrs methods. Also, this is a bit more efficient.
```

Reply:
hmm I think I meant "explicitly"